src/main/java/hudson/matrix/MatrixBuild.java

```diff
@@ -350,15 +350,26 @@ protected Result doRun(BuildListener listener) throws Exception {
             MatrixProject p = getProject();
             PrintStream logger = listener.getLogger();
 
-            // give axes a chance to rebuild themselves
-            activeConfigurations = p.rebuildConfigurations(this);
+            try {
+                // lock the project
+                p.buildLock();
+
+                // give axes a chance to rebuild themselves
+                activeConfigurations = p.rebuildConfigurations(this);
+
+                // list up aggregators
+                listUpAggregators(p.getPublishers().values());
+                listUpAggregators(p.getProperties().values());
+                listUpAggregators(p.getBuildWrappers().values());
 
-            // list up aggregators
-            listUpAggregators(p.getPublishers().values());
-            listUpAggregators(p.getProperties().values());
-            listUpAggregators(p.getBuildWrappers().values());
+                // deep copy the axes
+                axes = (AxisList) Jenkins.XSTREAM.fromXML(Jenkins.XSTREAM.toXML(p.getAxes()));
```

Comment:
> Yes that is the problem. The axisValues are cached and will be regenerated by rebuildConfigurations.

You are right. This seems to be specific to dynamic axis plugin (and your reproducer). I see `Axis#rebuild` is ill-designed in this regard, should return a copy of whole axis so deep copying can be implemented. I see no better way of doing so.